src/cancel_stream.rs

```diff
@@ -0,0 +1,293 @@
+use borsh::BorshSerialize;
+use solana_program::{
+    borsh as solana_borsh,
+    entrypoint::ProgramResult,
+    msg,
+    program::{invoke, invoke_signed},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar::{clock::Clock, Sysvar},
+};
+use spl_associated_token_account::create_associated_token_account;
+
+use crate::{
+    error::SfError,
+    state::{InstructionAccounts, TokenStreamData},
+    stream_safety::{initialized_account_sanity_check, metadata_sanity_check},
+    utils::{calculate_available, encode_base10, unpack_mint_account},
+};
+
+/// Cancel an SPL Token stream
+///
+/// The function will read the instructions from the metadata account and see
+/// if there are any unlocked funds. If so, they will be transferred to the
+/// stream recipient.
+pub(crate) fn cancel(program_id: &Pubkey, acc: InstructionAccounts) -> ProgramResult {
+    msg!("Cancelling SPL token stream");
+
+    let now = Clock::get()?.unix_timestamp as u64;
+    let mint_info = unpack_mint_account(&acc.mint)?;
+
+    // Sanity checks
+    initialized_account_sanity_check(program_id, acc.clone())?;
+    metadata_sanity_check(acc.clone())?;
+
+    let mut data = acc.metadata.try_borrow_mut_data()?;
+    let mut metadata: TokenStreamData = match solana_borsh::try_from_slice_unchecked(&data) {
+        Ok(v) => v,
+        Err(_) => return Err(SfError::InvalidMetadata.into()),
+    };
+
+    // If stream is expired, anyone can close it
+    if now < metadata.closable_at {
+        msg!("Stream not yet expired, checking authorization");
+        if acc.authority.key != acc.sender.key {
+            return Err(ProgramError::InvalidAccountData)
+        }
+
+        if !acc.authority.is_signer {
+            return Err(ProgramError::MissingRequiredSignature)
+        }
+    }
+
+    let recipient_available = calculate_available(
+        now,
+        metadata.ix.clone(),
+        metadata.ix.deposited_amount,
+        metadata.withdrawn_amount,
+    );
+
+    let streamflow_available = calculate_available(
+        now,
+        metadata.ix.clone(),
+        metadata.streamflow_fee_total,
+        metadata.streamflow_fee_withdrawn,
+    );
+
+    let partner_available = calculate_available(
```

Comment:
maybe move calculate_available to be a member of TokenStreamData since all the data needed to calculate it is member of TokenStreamData already